[RISCV] Add initial assembler/MC layer support for big-endian

djtodoro opened the pull request:
This patch adds basic assembler and MC layer infrastructure for
RISC-V big-endian targets (riscv32be/riscv64be):
      - Register big-endian targets in RISCVTargetMachine
      - Add big-endian data layout strings
      - Implement endianness-aware fixup application in assembler
        backend
      - Add byte swapping for data fixups on BE cores
      - Update MC layer components (AsmInfo, MCTargetDesc, Disassembler,
        AsmParser)
    
This provides the foundation for BE support but does not yet include:
      - Codegen patterns for BE
      - Load/store instruction handling
      - BE-specific subtarget features

djtodoro committed:
[RISCV] Add initial assembler/MC layer support for big-endian

This patch adds basic assembler and MC layer infrastructure for
RISC-V big-endian targets (riscv32be/riscv64be):

  - Register big-endian targets in RISCVTargetMachine
  - Add big-endian data layout strings
  - Implement endianness-aware fixup application in assembler
    backend
  - Add byte swapping for data fixups on BE cores
  - Update MC layer components (AsmInfo, MCTargetDesc, Disassembler,
    AsmParser)

This provides the foundation for BE support but does not yet include:
  - Codegen patterns for BE
  - Load/store instruction handling
  - BE-specific subtarget features

djtodoro commented:
A PR in psABI repo is here: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/470.

lenary commented:
This seems to only implement the first bullet point in your commit message, and assembler/relocation/fixup support, right? Please update the PR description.

I presume there will be follow-ups for the codegen/load-store support, plus subtarget info.

jh7370 commented on llvm/include/llvm/TargetParser/Triple.h:
```diff
@@ -48,62 +48,64 @@ class Triple {
-    lanai,          // Lanai: Lanai 32-bit
-    wasm32,         // WebAssembly with 32-bit pointers
-    wasm64,         // WebAssembly with 64-bit pointers
+    arm,         // ARM (little endian): arm, armv.*, xscale
```
This reformatting looks off to me. I'd expect the comments to line up the whole way down, but the renderscript comment and onwards is not at the same column.

djtodoro replied:
addressed in https://github.com/llvm/llvm-project/pull/146913

jh7370 commented on llvm/include/llvm/TargetParser/Triple.h:
```diff
@@ -48,62 +48,64 @@ class Triple {
+    ppc64le,     // PPC64LE: powerpc64le
+    r600,        // R600: AMD GPUs HD2XXX - HD6XXX
+    amdgcn,      // AMDGCN: AMD GCN GPUs
+    riscv32,     // RISC-V (32-bit): riscv32
+    riscv64,     // RISC-V (64-bit): riscv64
```
Worth specifying little endian here, to contrast with the big endian comment?

djtodoro replied:
I agree, addressed in https://github.com/llvm/llvm-project/pull/146913

jh7370 reviewed (commented):
llvm-objcopy and libObject changes look fine to me. If you split the llvm-objcopy changes into a separate PR, I'd be happy to approve directly.

wangpc-pp reviewed (commented):
How are we going to verify the correctness? Does qemu support big-endian riscv64?

djtodoro commented:
> How are we going to verify the correctness? Does qemu support big-endian riscv64?

Good question. We have an initial support for it developed internally inside MIPS, for both qemu and Linux kernel, but team from Codethink is trying to upstream similar effort already, so please check it here https://gitlab.com/CodethinkLabs/riscv_bigendian.

I am not sure how we merged Big Endian support for RISC-V in GNU GCC, but we should have agreed on convention first I think (effort initiated at https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/470.)

djtodoro commented:
> What is your intention here around the LLVM 21 release cycle, knowing this is a big feature and we are ~2 weeks from the branch date

Hey @lenary, thank for mentioning that. I agree it is a big feature, so no need to rush to include it in the release.

lenary commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -880,6 +882,34 @@ bool RISCVAsmBackend::addReloc(const MCFragment &F, const MCFixup &Fixup,
+  case FK_Data_8:
+    return true;
+
+  case RISCV::fixup_riscv_hi20:
+  case RISCV::fixup_riscv_lo12_i:
+  case RISCV::fixup_riscv_lo12_s:
+  case RISCV::fixup_riscv_pcrel_hi20:
+  case RISCV::fixup_riscv_pcrel_lo12_i:
+  case RISCV::fixup_riscv_pcrel_lo12_s:
+  case RISCV::fixup_riscv_jal:
+  case RISCV::fixup_riscv_branch:
+  case RISCV::fixup_riscv_call:
+  case RISCV::fixup_riscv_call_plt:
+  case RISCV::fixup_riscv_rvc_jump:
+  case RISCV::fixup_riscv_rvc_branch:
+    return false;
```
You're missing a bunch of instruction fixups here:

- `fixup_riscv_12_i`
- `fixup_riscv_rvc_imm`
- `fixup_riscv_qc_e_branch`
- `fixup_riscv_qc_e_32`
- `fixup_riscv_qc_abs20_u`
- `fixup_riscv_qc_e_call_plt`
- `fixup_riscv_nds_branch_10`

lenary commented on llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:
```diff
@@ -496,7 +496,8 @@ static ShadowMapping getShadowMapping(const Triple &TargetTriple, int LongSize,
                    TargetTriple.getArch() == Triple::aarch64_be;
   bool IsLoongArch64 = TargetTriple.isLoongArch64();
-  bool IsRISCV64 = TargetTriple.getArch() == Triple::riscv64;
+  bool IsRISCV64 = TargetTriple.getArch() == Triple::riscv64 ||
+                   TargetTriple.getArch() == Triple::riscv64be;
```
Given you're not aiming to do codegen yet, maybe this change should wait for a later patch when it can be tested?

lenary reviewed (commented):
The code mostly looks right, but it's going to need tests 😄

djtodoro committed:
[RISCV] Add tests for BE

lenary reviewed (approved):
LGTM! Thanks!

jrtc27 commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -892,8 +929,11 @@ void RISCVAsmBackend::applyFixup(const MCFragment &F, const MCFixup &Fixup,
   // For each byte of the fragment that the fixup touches, mask in the
   // bits from the fixup value.
+  // For big endian cores, data fixup should be swapped.
+  bool SwapValue = (Endian == llvm::endianness::big) && isDataFixup(Kind);
```
```suggestion
  bool SwapValue = Endian == llvm::endianness::big && isDataFixup(Kind);
```

jrtc27 commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
```diff
@@ -150,21 +152,23 @@ extern "C" LLVM_ABI LLVM_EXTERNAL_VISIBILITY void LLVMInitializeRISCVTarget() {
 
-    return "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128";
+  if (TT.isArch64Bit()) {
+    DL = (Twine(IsLittle ? "e" : "E") + "-m:e-p:64:64-i64:64-i128:128-n32:64-" +
```
This is quite a significant refactor. If you're going to decompose it like this, at least split it out properly? That is, pick e vs E in one place, pick 32-bit vs 64-bit in another and pick RVE vs RVI ABI in another, then glue it all together?

djtodoro replied:
Yes, I agree, thank you for suggestion.

djtodoro committed:
Address comments

djtodoro commented:
@jrtc27 Please let me know if this is okay now.

djtodoro commented on llvm/cmake/config.guess:
```diff
@@ -1003,7 +1003,7 @@ EOF
 	echo powerpcle-unknown-linux-gnu
 	exit ;;
-    riscv32:Linux:*:* | riscv64:Linux:*:*)
+    riscv32:Linux:*:* | riscv64:Linux:*:* | riscv32be:Linux:*:* | riscv64be:Linux:*:*)
```
Okay, I will remove it from this change.

djtodoro committed:
Address comments

- Drop change in llvm/cmake/config.guess
- Align computeDataLayout with X86 and ARM

topperc reviewed (approved):
LGTM

jrtc27 commented:
Do we need to gate this behind -menable-experimental-extensions (or an equivalent given it's not really experimental)? There is no ABI specification for it yet and I would be concerned about future incompatibility. GCC implements *an* ABI but IIRC there were concerns about whether it was actually a sensible ABI or organically came into being, possibly around things like structs passed in register pairs?

jrtc27 commented:
See the thread starting at https://github.com/riscv-non-isa/riscv-elf-psabi-doc/issues/265#issuecomment-2148592299

djtodoro committed:
Apply refactor

MaskRay commented on llvm/test/MC/RISCV/bigendian-fixups.s:
```diff
@@ -0,0 +1,60 @@
+# RUN: llvm-mc -filetype=obj -triple=riscv32be < %s \
+# RUN:     | llvm-objdump --no-print-imm-hex -M no-aliases -d - \
```
Test inline relocations with `llvm-objdump -dr -`. It's important to test the encoding for at least a few insns, so you should omit --no-print-imm-hex

djtodoro committed:
Improve tests

MaskRay commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -869,6 +871,41 @@ bool RISCVAsmBackend::addReloc(const MCFragment &F, const MCFixup &Fixup,
+  case FK_Data_8:
+    return true;
+
+  case RISCV::fixup_riscv_hi20:
```
delete all `fixup_riscv_*` cases - these are all instruction kinds.
If we ever have fixups for data relocations, we could add them like FK_Data_8 in the future.

djtodoro committed:
Remove fixup_riscv_*

djtodoro committed:
Apply clang-format

MaskRay commented:
Description:

> - Handle BE targets in AddressSanitizer

While this might be on the roadmap, I don't see it in the patch.

djtodoro committed:
Drop BE support from TargetLoweringObjectFileImpl

lenary reviewed (approved):
I'm happy with this as is.

I agree for the need for some kind of "experimental" warning, but I think that's better to come from clang, probably in `RISCVTargetInfo::handleTargetFeatures` - where we also have things like the `+experimental` feature so it can be disabled if need be.

I think this is safe to land as there have been no changes to clang, clang assumes little-endian for RISC-V, and most users will be using clang rather than e.g. `llvm-mc`.

djtodoro commented on llvm/lib/MC/MCAsmStreamer.cpp:
```diff
@@ -2341,6 +2341,9 @@ void MCAsmStreamer::AddEncodingComment(const MCInst &Inst,
   getAssembler().getEmitter().encodeInstruction(Inst, Code, Fixups, STI);
 
+  // RISC-V instructions are always little-endian, even on BE systems.
+  bool ForceLE = getContext().getTargetTriple().isRISCV();
```
Yes, noted, I agree.

djtodoro commented:
> I'm happy with this as is.
> 
> I agree for the need for some kind of "experimental" warning, but I think that's better to come from clang, probably in `RISCVTargetInfo::handleTargetFeatures` - where we also have things like the `+experimental` feature so it can be disabled if need be.
> 

It does make sense. I will address it along with `clang` patches.

> I think this is safe to land as there have been no changes to clang, clang assumes little-endian for RISC-V, and most users will be using clang rather than e.g. `llvm-mc`.

Yes, I agree.

@lenary Thanks a lot!
